feat(launchpad): open in IDE modal and feature

lmiller1990 opened the pull request:
## Features

- Configure preferred IDE from launchpad and app if one isn't configured
- Open file in preferred IDE
- Open project directory in preferred IDE
- Open project directory in finder

## Testing

See videos - they show the feature, where it is, how it works. Be good to get someone else to give it a test, too. I also added a bunch of e2e tests.

## More Info
I changed up how reporter communicates w/ event manager and the back end when opening files. Reporter is now less smart - it just emits an event, and the runner will handle the request to open a file via GraphQL. This made the React -> Vue -> event manager interop much more simple and avoided a bunch of hacks, as well as made the types better, and allowed for better e2e tests.

Prevously, the front-end passed the preferred editor with the "open in file" request, which I don't think is ideal. We should only have one source of truth for the preferred editor, and that should be on the server, not in the front-end.

I left in the existing "open in file" functionality since removing or changing it breaks a lot of system tests.

## Demos

Demo: set IDE and open from app

https://user-images.githubusercontent.com/19196536/142591349-38554b7b-0a8a-4513-826b-b0bd98725490.mov

Demo: set IDE and open from launchpad

https://user-images.githubusercontent.com/19196536/142591608-71a01a76-bcf5-47b5-83d5-3d6dc77de71e.mov

lmiller1990 commented on packages/app/src/pages/Runner.vue:
```diff
@@ -1,25 +1,37 @@
 
 const query = useQuery({ query: SpecPageContainerDocument })
+
+export type GqlWithCurrentProject = SpecPageContainerQuery & {
```
This feels hacky, @tgriesser can we sync on this? Also notified in [slack](https://cypressio.slack.com/archives/C027U3ZEJ6Q/p1637307934004600)

elevatebart reviewed (changes requested):
I wonder if it was intended but when testing this feature, I found an issue.

If you set your editor in the settings page using the drop-down, wait a second and then refresh the page, the value is empty again.

I will look into why.

elevatebart commented on packages/frontend-shared/src/gql-components/ChooseExternalEditor.vue:
```diff
@@ -0,0 +1,164 @@
+import { useI18n } from '@cy/i18n'
+import Select from '@cy/components/Select.vue'
+import Input from '@cy/components/Input.vue'
+import VSCode from '~icons/logos/visual-studio-code'
+import Atom from '~icons/logos/atom-icon'
+import Webstorm from '~icons/logos/webstorm'
+import Vim from '~icons/logos/vim'
+import Sublime from '~icons/logos/sublimetext-icon'
+import Emacs from '~icons/logos/emacs'
+import IconTerminal from '~icons/mdi/terminal'
```
Why are those icons not part of the `env-editors.ts` ?

elevatebart replied:
I get it, because they are not available there.

Maybe we could use import meta to import a glob of logos and only reference logos by name on the server...

This is an architecture concern though.

elevatebart replied:
Could we at least add one icon corresponding to `computer` for the open in file system choice.

And use the terminal icon from cypress instead of MDI?

lmiller1990 replied:
Done and done for icons, let's consider the Vite globbing, too, at some point...

![image](https://user-images.githubusercontent.com/19196536/142793578-7c71bfd2-72b8-405f-8fc7-02f9db78b5f3.png)

elevatebart replied:
love it

lmiller1990 commented:
@elevatebart 

> I wonder if it was intended but when testing this feature, I found an issue.

I made a change in a commit and looks like I didn't test it - setting an editor on that page was completely not working as it should have been. Huge oversight on my end, extended the e2e test to make sure it does not regress: https://github.com/cypress-io/cypress/pull/18975/commits/4ec6bbfa00379f56d1a891403a22977945e12c3f

elevatebart reviewed (dismissed):
Except for some icons not aligned it works

elevatebart commented on packages/frontend-shared/cypress/e2e/support/e2eProjectDirs.ts:
```diff
@@ -65,6 +65,7 @@ export const e2eProjectDirs = [
   'screen-size',
   'shadow-dom-global-inclusion',
   'spec-generation',
+  'spec-name-special-characters',
```
Why is this file updated?
@tgriesser should we gitignore it?

ZachJW34 reviewed (dismissed):
Tested, looks good. I'm curious about the comment you made about the hacky GraphQL types. Can we not just use the normal fragment types here? Might require some refactoring of the fragments though

elevatebart commented on packages/data-context/src/actions/FileActions.ts:
```diff
@@ -35,4 +37,28 @@ export class FileActions {
+  openFile (absolute: string, line: number = 1, column: number = 1) {
+    const binary = this.ctx.coreData.localSettings.preferences.preferredEditorBinary
+
+    if (!binary || !absolute) {
+      this.ctx.debug('cannot open file without binary')
+
+      return
+    }
+
+    if (binary === 'computer') {
+      try {
+        this.ctx.electronApi.showItemInFolder(absolute)
+      } catch (err) {
+        this.ctx.debug('error opening file: %s', (err as Error).stack)
+      }
+
+      return
+    }
+
+    launchEditor(`${absolute}:${line}:${column}`, `"${binary}"`, (__: unknown, errMsg: string) => {
+      this.ctx.debug('error opening file: %s', errMsg)
+    })
```
Do you think this implementation should be done here or could it be done in another file?

lmiller1990 replied:
I guess, why? Where?

lmiller1990 commented:
@ZachJW34 I summed up about the types here: https://cypressio.slack.com/archives/C027U3ZEJ6Q/p1637307934004600 basically in GraphQL `currentProject` is nullable, but on this page it will never be null - I want some way to express that.

For now I reverted the hack and just added some null checks. Long term we will wait for the `!` operator on fields to land in urql: https://github.com/FormidableLabs/urql/pull/2018

Then we can do

```
query Q {
  currentProject!
```

... represents that `currentProject` cannot be null.
